OSDOCS-21326:Support of BYO firewall workflow for OSD-GCP deployments. (#118155)
openshift/openshift-docs: modules/osd-gcp-firewall-rules.adoc

// Module included in the following assemblies:
//
// * osd_planning/gcp-ccs.adoc

:_mod-docs-content-type: PROCEDURE
[id="osd-gcp-firewall-rules_{context}"]
= Create firewall rules for reduced Workload Identity Federation permissions on {gcp-short}

[role="_abstract"]

> Starting with {product-title} version 4.21, the `osd_deployer` and `openshift_gcp_ccm` Workload Identity Federation (WIF) custom roles no longer include `compute.firewalls.create` and `compute.firewalls.delete` permissions. You must create the required firewall rules in your {gcp-short} project before installing a cluster. This approach allows your network security team to keep control over firewall rule management without granting broad firewall creation and deletion privileges to {product-title}-managed service accounts.

Reviewer: This section is slightly misleading at the moment, as we can not yet remove `compute.firewalls` permissions from the service project. The only permission reduction currently achievable with BYO firewall rules is for the shared vpc deployment mode. I would omit this section or instead highlight the permission benefits for shared vpc instead.

[IMPORTANT]
====
If you create firewall rules by using this procedure, you are responsible for managing firewall rules through the lifetime of the cluster. The cluster's Cloud Controller Manager does not create or delete firewall rules automatically. If you expose a new service by using a load balancer or change the ingress configuration, you must create or update the corresponding firewall rules manually.
====

.Prerequisites

* You have a {gcp-short} project configured for {product-title} deployment.
* You have downloaded the latest version of the {cluster-manager} CLI (`ocm`) for your operating system from the link:https://console.redhat.com/openshift/downloads[Downloads] page on {cluster-manager}.
* You have a VPC network created in the project for the cluster to use. For more information, see link:https://cloud.google.com/vpc/docs/create-modify-vpc-networks[Create and manage VPC networks].
* You have created a Workload Identity Federation (WIF) configuration and know the WIF configuration ID. For more information, see _Create a Workload Identity Federation configuration_ in the _Additional resources_ section.

.Procedure

> . Create the firewall rules by running the `ocm gcp create firewall-rules` command in `manual` or `automatic` mode:

Reviewer: ```suggestion
. Create the firewall rules by running the `ocm gcp create firewall-rules` command in `manual` mode:
```
At this time, only manual mode is supported. The user is expected to verify and run the generated `apply.sh` script. Please make a note of this in the documentation.

+
--
* In `manual` mode, the command generates a shell script containing the required `gcloud` commands. You must review and run the script yourself.

> * In `automatic` mode, the command creates the firewall rules directly in your {gcp-short} project.

Reviewer: ```suggestion
* In `manual` mode, the command generates a shell script containing the required `gcloud` commands. You must review and run the script yourself.
```

Reviewer: https://github.com/openshift/openshift-docs/pull/118155#discussion_r3874267694

--
+
** To create the firewall rules in manual mode, run the following command:
+
[source,terminal]
----
$ ocm gcp create firewall-rules \
    --mode manual \
    --name <firewall_rule_prefix> \
    --wif-config <wif_config_id> \
    --project-id <host_project_id> \
    --vpc-name <vpc_name> \
    --machine-cidr <machine_cidr> \
    --profile <public|private> \
    --output-file <output_file_path>
----
+
** To create the firewall rules in automatic mode, run the following command:
+
[source,terminal]
----
$ ocm gcp create firewall-rules \
    --mode automatic \
    --name <firewall_rule_prefix> \
    --wif-config <wif_config_id> \
    --project-id <host_project_id> \
    --vpc-name <vpc_name> \
    --machine-cidr <machine_cidr> \
    --profile <public|private>

> ----

Reviewer: ```suggestion
```

Reviewer: https://github.com/openshift/openshift-docs/pull/118155#discussion_r3874267694

. If you used manual mode, review the generated script and run it:
+
[source,terminal]
----
$ chmod +x <output_file_path>
$ ./<output_file_path>
----

.Verification

* Verify that the firewall rules were created successfully:
+
[source,terminal]
----
$ gcloud compute firewall-rules list \
  --project=<host_project_id> \
  --filter="network=<vpc_name>" \
  --format="table(name, direction, allowed, sourceRanges.list():label=SRC_RANGES, targetServiceAccounts.list():label=TARGET_SA)"
----

[IMPORTANT]
====
Because the Cloud Controller Manager does not have permissions to create or delete firewall rules, the cluster emits events when a load balancer service requires a new firewall rule. You must check for these events and create the required firewall rules manually by running the following command:

[source,terminal]
----
$ oc get events -n openshift-ingress --field-selector="reason=LoadBalancerManualChange"
----

The event output includes the `gcloud` command to create the required firewall rule. Run the provided command to allow traffic to the load balancer service. If you delete a load balancer service, delete the corresponding firewall rule manually.

> ====

Reviewer: ```suggestion
```

Omit for now. See previous comment.

Reviewer: https://github.com/openshift/openshift-docs/pull/118155#discussion_r3874255876

//
// * osd_planning/gcp-ccs.adoc

:_mod-docs-content-type: PROCEDURE
[id="osd-gcp-firewall-rules_{context}"]
= Create firewall rules for reduced Workload Identity Federation permissions on {gcp-short}

[role="_abstract"]
Some {product-title} on {GCP} clusters use a reduced set of permissions for the Workload Identity Federation (WIF) configuration. To allow the cluster to function properly, you must create firewall rules in your {gcp-short} project that allow traffic between the cluster's control plane and worker nodes. This approach allows your network security team to maintain control over firewall rule management in the shared VPC host project without granting broad firewall creation and deletion privileges to {product-title}-managed service accounts.

[IMPORTANT]
====
If you create firewall rules by using this procedure, you are responsible for managing firewall rules through the lifetime of the cluster. The cluster's Cloud Controller Manager does not create or delete firewall rules automatically. If you expose a new service by using a load balancer or change the ingress configuration, you must create or update the corresponding firewall rules manually.
====

.Prerequisites

> 

Reviewer 2: Optional. 

```suggestion
Ensure you have done the following tasks: 
```

> * You have a {gcp-short} project configured for {product-title} deployment.

Reviewer 2: ```suggestion
* Configure a {gcp-short} project for {product-title} deployment.
```

> * You have downloaded the latest version of the {cluster-manager} CLI (`ocm`) for your operating system from the link:https://console.redhat.com/openshift/downloads[Downloads] page on {cluster-manager}.

Reviewer 2: ```suggestion
* Download the latest version of the {cluster-manager} CLI (`ocm`) for your operating system from the link:https://console.redhat.com/openshift/downloads[Downloads] page on {cluster-manager}.
```

* You have a VPC network created in the project for the cluster to use. For more information, see link:https://cloud.google.com/vpc/docs/create-modify-vpc-networks[Create and manage VPC networks].

> * You have created a Workload Identity Federation (WIF) configuration and know the WIF configuration ID. For more information, see _Create a Workload Identity Federation configuration_ in the _Additional resources_ section.

Reviewer 2: ```suggestion
* Create a Workload Identity Federation (WIF) configuration and know the WIF configuration ID. For more information, see _Create a Workload Identity Federation configuration_ in the _Additional resources_ section.
```
